core/app/models/spree/stock_location.rb

```diff
@@ -51,11 +51,14 @@ def set_up_stock_item(variant)
 
     # Returns an instance of StockItem for the variant id.
     #
-    # @param variant_id [String] The id of a variant.
+    # @param variant_or_id [String] The id of a variant.
     #
     # @return [StockItem] Corresponding StockItem for the StockLocation's variant.
-    def stock_item(variant_id)
-      stock_items.where(variant_id: variant_id).order(:id).first
+    def stock_item(variant_or_id)
+      stock_items.detect do |stock_item|
```

Comment:
👋🏼 @mamhoff I am curious, have we done any real world benchmarking for how this is going to impact stores with a large number of variants/stock items. I work on a store which has ~90K stock items in each stock location and an isolated call to this method is _much_ slower in Ruby due to the need to load all records in memory before iterating over them. The benchmark below is just from my local development environment so doesn't account for network traffic/database performance etc. This is also a "worst case" scenario where the stock item is the last one in the list (based on the order the db returns them).
```ruby
[1] pry(main)> Benchmark.measure { Spree::StockLocation.first.stock_item(93917) }
=> #<Benchmark::Tms:0x00007fc663c98098 @cstime=0.0, @cutime=0.0, @label="", @real=0.2266759998165071, @stime=0.026478, @total=0.07949200000000012, @utime=0.05301400000000012>
[2] pry(main)> Benchmark.measure { Spree::StockLocation.first.stock_items.detect { |i| i.variant_id == 93917 } }
=> #<Benchmark::Tms:0x00007fc64944ccf0 @cstime=0.0, @cutime=0.0, @label="", @real=2.5800430001690984, @stime=0.06728000000000023, @total=2.5387829999999996, @utime=2.4715029999999993>
```
I realize this example isn't indicative of real world usage, that's why I am curious if you've seen a performance increase with the ruby code when adding variants etc? An alternative here could be to use lazy enumeration to save us from having to load all records in memory 🤔

Reply:
Good point; I came to this by going through all the callbacks on `Spree::LineItem` and found `Spree::OrderInventory` to make heavy use of this method. We have some 4000 stock items per stock location, so it's not that bad here, but we also have really large orders with some 200 line items; and we provide bulk adding / removing items to those orders, so anything that creates n+1 queries on line items really hurts us. 

My instinct so far has been that the combination of preloading records and iterating is often a lot faster than even 10 DB queries, especially when looking at things in production, and it's gotten me really really far with getting some of our more expensive endpoints faster. 

I guess this is a question of Solidus policy: We do want to serve high-traffic sites, but do we also want to serve super-large-catalog sites? What are the trade-offs? :)

Reply:
> We do want to serve high-traffic sites, but do we also want to serve super-large-catalog sites? 

I think whenever possible we should try to do both 😸 I am 💯 with you on this work to reduce DB queries by operating on in-memory preloaded records, it will improve performance significantly!

I am curious if in this case we can invert the lookup by loading all stock items for a variant and looking for one with a matching stock location where this is called from - 
https://github.com/solidusio/solidus/blob/master/core/app/models/spree/order_inventory.rb#L65 🤔 It is less likely that a store has tens of thousands of stock locations, whereas that is a totally reasonable number of stock items imho.

Reply:
https://github.com/solidusio/solidus/pull/4303